File: node/lib/upload_dir.mjs
import fs from "node:fs";
import path from "node:path";
import { apiPost, extractId } from "./api.mjs";
import { uploadImageForDocument } from "./attachment.mjs";
import { printJson } from "./utils.mjs";

const DEFAULT_IGNORES = [
  ".*", "__pycache__", "node_modules", ".git", ".svn", ".DS_Store", "Thumbs.db",
];
const INDEX_FILENAMES = new Set(["readme.md", "_index.md", "index.md"]);

function globToRegex(glob) {
  let re = "^";
  for (const c of glob) {
    if (c === "*") re += ".*";
    else if (c === "?") re += ".";
    else if (".+^$()|[]{}\\".includes(c)) re += "\\" + c;
    else re += c;
  }
  return new RegExp(re + "$");
}

function shouldIgnore(name, patterns) {
  return patterns.some((p) => globToRegex(p).test(name));
}

function stripFrontmatter(text) {
  if (!text.startsWith("---")) return { title: null, body: text };
  const end = text.indexOf("\n---", 3);
  if (end === -1) return { title: null, body: text };
  const block = text.slice(3, end);
  const body = text.slice(end + 4).replace(/^\n+/, "");
  let title = null;
  for (const line of block.split(/\r?\n/)) {
    const t = line.trim();
    if (t.toLowerCase().startsWith("title:")) {
      title = t.slice(6).trim().replace(/^["']|["']$/g, "");
      break;
    }
  }
  return { title, body };
}

async function processMarkdownImages(text, documentId, baseDir) {
  const mdRe = /!\[([^\]]*)\]\(([^)]+)\)/g;
  const htmlRe = /(<img[^>]+src=["'])([^"']+)(["'][^>]*>)/g;

  async function resolveUrl(raw) {
    const parts = raw.trim().split(/\s+/);
    let clean;
    try { clean = decodeURIComponent(parts[0]); } catch { clean = parts[0]; }
    if (clean.startsWith("http") || clean.startsWith("/api/attachments") || clean.startsWith("data:")) return null;
    const localPath = path.isAbsolute(clean) ? clean : path.normalize(path.join(baseDir, clean));
    if (!fs.existsSync(localPath)) return null;
    const newUrl = await uploadImageForDocument(localPath, documentId);
    if (!newUrl) return null;
    return parts.length > 1 ? `${newUrl} ${parts.slice(1).join(" ")}` : newUrl;
  }

  async function replaceAll(source, re, build) {
    const matches = [];
    let m;
    re.lastIndex = 0;
    while ((m = re.exec(source)) !== null) matches.push({ m, index: m.index });
    let out = source;
    for (let i = matches.length - 1; i >= 0; i--) {
      const { m, index } = matches[i];
      const replacement = await build(m);
      if (replacement !== null) {
        out = out.slice(0, index) + replacement + out.slice(index + m[0].length);
      }
    }
    return out;
  }

  text = await replaceAll(text, mdRe, async (m) => {
    const u = await resolveUrl(m[2]);
    return u ? `![${m[1]}](${u})` : null;
  });
  text = await replaceAll(text, htmlRe, async (m) => {
    const u = await resolveUrl(m[2]);
    return u ? `${m[1]}${u}${m[3]}` : null;
  });
  return text;
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

export async function handleUploadDir(_action, f) {
  const targetDir = f.dir;
  if (!targetDir || !fs.existsSync(targetDir) || !fs.statSync(targetDir).isDirectory()) {
    printJson({ ok: false, error: `Directory not found: ${targetDir}` });
    process.exit(1);
  }
  if (!f.collectionId) { printJson({ ok: false, error: "需要 --collection-id" }); process.exit(1); }
  const collectionId = extractId(f.collectionId, "collection");
  const delay = f.delay ? parseFloat(f.delay) : 1.0;
  const dryRun = !!f.dryRun;
  const ignorePatterns = [...DEFAULT_IGNORES, ...(f.ignore || [])];
  const actions = [];

  async function createDoc(title, text, parentId) {
    if (dryRun) return { ok: true, data: { id: `<dry:${title}>` } };
    const payload = { collectionId, title, text, publish: true };
    if (parentId && !String(parentId).startsWith("<dry:")) payload.parentDocumentId = parentId;
    const res = await apiPost("documents.create", payload);
    await sleep(delay * 1000);
    return res;
  }

  async function updateDoc(id, title, text) {
    if (dryRun || String(id).startsWith("<dry:")) return { ok: true };
    const res = await apiPost("documents.update", { id, title, text });
    await sleep(delay * 1000);
    return res;
  }

  async function processDir(currentDir, parentDocId = null) {
    const entries = (await fs.promises.readdir(currentDir)).sort();
    let indexFile = null;
    const otherMd = [];
    const subdirs = [];
    for (const name of entries) {
      if (shouldIgnore(name, ignorePatterns)) continue;
      const full = path.join(currentDir, name);
      const stat = await fs.promises.stat(full);
      if (stat.isDirectory()) subdirs.push(name);
      else if (stat.isFile() && name.toLowerCase().endsWith(".md")) {
        if (INDEX_FILENAMES.has(name.toLowerCase()) && !indexFile) indexFile = name;
        else otherMd.push(name);
      }
    }

    for (const sub of subdirs) {
      const full = path.join(currentDir, sub);
      let subIndexPath = null;
      for (const cand of INDEX_FILENAMES) {
        const p = path.join(full, cand);
        if (fs.existsSync(p)) { subIndexPath = p; break; }
      }
      let folderTitle, folderText;
      if (subIndexPath) {
        const raw = await fs.promises.readFile(subIndexPath, "utf8");
        const { title, body } = stripFrontmatter(raw);
        folderTitle = title || sub;
        folderText = body;
      } else {
        folderTitle = sub;
        folderText = `# ${folderTitle}\n\n本目录索引页。`;
      }
      const res = await createDoc(folderTitle, folderText, parentDocId);
      if (res.ok) {
        const folderDocId = res.data.id;
        actions.push({ type: "folder", title: folderTitle, id: folderDocId, status: "ok" });
        console.log(`[folder] ${folderTitle}  → ${folderDocId}`);
        if (subIndexPath && !dryRun) {
          const newText = await processMarkdownImages(folderText, folderDocId, full);
          if (newText !== folderText) await updateDoc(folderDocId, folderTitle, newText);
        }
        await processDir(full, folderDocId);
      } else {
        actions.push({ type: "folder", title: folderTitle, status: "fail", error: res });
        console.log(`[folder FAIL] ${folderTitle}: ${JSON.stringify(res)}`);
      }
    }

    for (const entry of otherMd) {
      const full = path.join(currentDir, entry);
      const content = await fs.promises.readFile(full, "utf8");
      const { title: fmTitle, body } = stripFrontmatter(content);
      const title = fmTitle || path.basename(entry, path.extname(entry));
      const res = await createDoc(title, body, parentDocId);
      if (res.ok) {
        const docId = res.data.id;
        actions.push({ type: "document", title, id: docId, status: "ok" });
        console.log(`[doc] ${title}  → ${docId}`);
        if (!dryRun) {
          const newText = await processMarkdownImages(body, docId, currentDir);
          if (newText !== body) await updateDoc(docId, title, newText);
        }
      } else {
        actions.push({ type: "document", title, status: "fail", error: res });
        console.log(`[doc FAIL] ${title}: ${JSON.stringify(res)}`);
      }
    }
  }

  console.log(`${dryRun ? "[DRY RUN] " : ""}Uploading ${targetDir} → collection ${collectionId}`);
  await processDir(targetDir);
  console.log();
  printJson({
    ok: true,
    dry_run: dryRun,
    total: actions.length,
    ok_count: actions.filter((a) => a.status === "ok").length,
    fail_count: actions.filter((a) => a.status === "fail").length,
  });
}

File: python/scripts/upload_dir.py
import os
import sys
import json
import time
import re
import fnmatch
from .api import api_post, extract_id, require_danger_confirmation
from .attachment import upload_image_for_document


# 默认忽略：隐藏文件、版本控制、依赖目录、系统垃圾、模板自身
DEFAULT_IGNORES = [
    ".*",
    "__pycache__",
    "node_modules",
    ".git",
    ".svn",
    ".DS_Store",
    "Thumbs.db",
]

# 被识别为"目录索引"的 markdown 文件名（小写比较）
INDEX_FILENAMES = {"readme.md", "_index.md", "index.md"}


def _should_ignore(name, patterns):
    return any(fnmatch.fnmatch(name, p) for p in patterns)


def _strip_frontmatter(text):
    """支持 --- title: xxx --- 的 YAML frontmatter，返回 (title_or_none, body_text)。"""
    if not text.startswith("---"):
        return None, text
    end = text.find("\n---", 3)
    if end == -1:
        return None, text
    block = text[3:end]
    body = text[end + 4:].lstrip("\n")
    title = None
    for line in block.splitlines():
        line = line.strip()
        if line.lower().startswith("title:"):
            title = line.split(":", 1)[1].strip().strip("\"'")
            break
    return title, body


def process_markdown_images(text, document_id, base_dir):
    """识别 markdown / HTML 图片中的本地路径，上传后替换为 Outline 链接。"""
    import urllib.parse

    md_pattern = r"!\[([^\]]*)\]\(([^)]+)\)"
    html_pattern = r'(<img[^>]+src=["\'])([^"\']+)(["\'][^>]*>)'

    def upload_and_get_url(image_url):
        parts = image_url.split(maxsplit=1)
        clean_url = urllib.parse.unquote(parts[0])
        if (
            not clean_url.startswith("http")
            and not clean_url.startswith("/api/attachments")
            and not clean_url.startswith("data:")
        ):
            local_path = clean_url if os.path.isabs(clean_url) else os.path.normpath(os.path.join(base_dir, clean_url))
            if os.path.exists(local_path):
                new_url = upload_image_for_document(local_path, document_id)
                if new_url:
                    return f"{new_url} {parts[1]}" if len(parts) > 1 else new_url
        return None

    def replace_md(m):
        new_url = upload_and_get_url(m.group(2).strip())
        return f"![{m.group(1)}]({new_url})" if new_url else m.group(0)

    def replace_html(m):
        new_url = upload_and_get_url(m.group(2).strip())
        return f"{m.group(1)}{new_url}{m.group(3)}" if new_url else m.group(0)

    text = re.sub(md_pattern, replace_md, text)
    text = re.sub(html_pattern, replace_html, text)
    return text


def handle_upload_dir(args):
    target_dir = args.dir
    collection_id = extract_id(args.collection_id, "collection")
    rate_limit_delay = args.delay
    dry_run = args.dry_run
    if not dry_run:
        require_danger_confirmation(
            "upload_dir 非 dry-run 批量上传",
            getattr(args, "confirm", False) or getattr(args, "confirm_dangerous", False),
        )

    if not os.path.isdir(target_dir):
        print(json.dumps({"ok": False, "error": f"Directory not found: {target_dir}"}, ensure_ascii=False))
        sys.exit(1)

    ignore_patterns = list(DEFAULT_IGNORES)
    if args.ignore:
        ignore_patterns.extend(args.ignore)

    actions = []  # for dry-run output / final summary

    def create_doc(title, text, parent_id=None):
        if dry_run:
            return {"ok": True, "data": {"id": f"<dry:{title}>"}}
        payload = {
            "collectionId": collection_id,
            "title": title,
            "text": text,
            "publish": True,
        }
        if parent_id and not parent_id.startswith("<dry:"):
            payload["parentDocumentId"] = parent_id
        res = api_post("documents.create", payload)
        time.sleep(rate_limit_delay)
        return res

    def update_doc(doc_id, title, text):
        if dry_run or doc_id.startswith("<dry:"):
            return {"ok": True}
        payload = {"id": doc_id, "title": title, "text": text}
        res = api_post("documents.update", payload)
        time.sleep(rate_limit_delay)
        return res

    def process_directory(current_dir, parent_doc_id=None):
        entries = sorted(os.listdir(current_dir))

        # 先识别目录索引文件（README.md / _index.md / index.md）
        index_file = None
        other_md_files = []
        subdirs = []
        for entry in entries:
            if _should_ignore(entry, ignore_patterns):
                continue
            full = os.path.join(current_dir, entry)
            if os.path.isdir(full):
                subdirs.append(entry)
            elif os.path.isfile(full) and entry.lower().endswith(".md"):
                if entry.lower() in INDEX_FILENAMES and index_file is None:
                    index_file = entry
                else:
                    other_md_files.append(entry)

        # 处理子目录：每个子目录创建一个父文档
        for sub in subdirs:
            full = os.path.join(current_dir, sub)
            # 子目录是否带索引文件？
            sub_index_path = None
            for cand in INDEX_FILENAMES:
                p = os.path.join(full, cand)
                if os.path.exists(p):
                    sub_index_path = p
                    break

            if sub_index_path:
                with open(sub_index_path, "r", encoding="utf-8") as f:
                    raw = f.read()
                fm_title, body = _strip_frontmatter(raw)
                folder_title = fm_title or sub
                folder_text = body
            else:
                folder_title = sub
                folder_text = f"# {folder_title}\n\n本目录索引页。"

            res = create_doc(folder_title, folder_text, parent_doc_id)
            if res.get("ok"):
                folder_doc_id = res["data"]["id"]
                actions.append({"type": "folder", "title": folder_title, "id": folder_doc_id, "status": "ok"})
                print(f"[folder] {folder_title}  → {folder_doc_id}")
                # 处理索引文件中的图片
                if sub_index_path and not dry_run:
                    new_text = process_markdown_images(folder_text, folder_doc_id, full)
                    if new_text != folder_text:
                        update_doc(folder_doc_id, folder_title, new_text)
                process_directory(full, folder_doc_id)
            else:
                actions.append({"type": "folder", "title": folder_title, "status": "fail", "error": res})
                print(f"[folder FAIL] {folder_title}: {res}")

        # 处理本目录的非索引 md 文件
        for entry in other_md_files:
            full = os.path.join(current_dir, entry)
            with open(full, "r", encoding="utf-8") as f:
                content = f.read()
            fm_title, body = _strip_frontmatter(content)
            title = fm_title or os.path.splitext(entry)[0]

            res = create_doc(title, body, parent_doc_id)
            if res.get("ok"):
                doc_id = res["data"]["id"]
                actions.append({"type": "document", "title": title, "id": doc_id, "status": "ok"})
                print(f"[doc] {title}  → {doc_id}")
                if not dry_run:
                    new_text = process_markdown_images(body, doc_id, current_dir)
                    if new_text != body:
                        update_doc(doc_id, title, new_text)
            else:
                actions.append({"type": "document", "title": title, "status": "fail", "error": res})
                print(f"[doc FAIL] {title}: {res}")

    print(f"{'[DRY RUN] ' if dry_run else ''}Uploading {target_dir} → collection {collection_id}")
    process_directory(target_dir)

    summary = {
        "ok": True,
        "dry_run": dry_run,
        "total": len(actions),
        "ok_count": sum(1 for a in actions if a.get("status") == "ok"),
        "fail_count": sum(1 for a in actions if a.get("status") == "fail"),
    }
    print("\n" + json.dumps(summary, indent=2, ensure_ascii=False))


def setup_upload_dir_parser(subparsers):
    p = subparsers.add_parser(
        "upload_dir",
        help="递归上传 markdown 目录到集合（保留层级；自动上传本地图片；支持 README/_index 作为目录索引）",
    )
    p.add_argument("--dir", required=True, help="本地目录路径")
    p.add_argument("--collection-id", required=True, help="目标集合 ID 或 URL")
    p.add_argument("--delay", type=float, default=1.0, help="每次 API 调用间隔秒（默认 1.0，遇 429 调大）")
    p.add_argument("--dry-run", action="store_true", help="只打印将要执行的操作，不写入 Outline")
    p.add_argument("--confirm", action="store_true", help="危险操作保护开启时，确认执行非 dry-run 批量上传")
    p.add_argument("--confirm-dangerous", action="store_true", help="同 --confirm；语义更明确，供脚本使用")
    p.add_argument("--ignore", action="append", help="额外忽略 glob 模式（可多次指定）")
    p.set_defaults(func=handle_upload_dir)
